### PR TITLE
[dashboard] Add Git Providers

### PR DESCRIPTION
**Pull request by AlexTugarev**
This is Part 1 of #3401

Screen Shot 2021-03-22 at 06 11 42

### CODE REVIEW

**Review by svenefftinge — APPROVED** [review]
Works nicely. The designs are still rough, but that can be done in another PR so we can merge this into dashboard-v2 and then merge that one into main.

File: components/dashboard/src/components/ContextMenu.tsx [hunk 1]
```diff
@@ -53,15 +56,15 @@ function ContextMenu(props: ContextMenuProps) {
...
+                    {enhancedEntries.map((e, index) => {
                         const clickable = e.href || e.onClick;
                         const entry = <div key={e.title} className={`px-4 flex py-3 ${clickable?'hover:bg-gray-200':''} text-sm leading-1 ${e.customFontStyle || font} ${e.separator? ' border-b border-gray-200':''}`} >
                             <div>{e.title}</div><div className="flex-1"></div>{e.active ? <div className="pl-1 font-semibold">&#x2713;</div>: null}
                         </div>
                         if (!clickable) {
                             return entry;
                         }
-                        return <a key={e.title} href={e.href} onClick={e.onClick}>
+                        return <a key={`entry-${menuId}-${index}-${e.title}`} href={e.href} onClick={e.onClick}>
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/dashboard/src/components/ContextMenu.tsx:67]: when rendering a table with rows which each contain a context menu with the same title you get the "unique" key problem.

File: components/dashboard/src/components/Modal.tsx [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { Disposable, DisposableCollection } from "@gitpod/gitpod-protocol";
 import { useEffect } from "react";
 
 export default function Modal(props: {
-    children: React.ReactChild[] | React.ReactChild,
+    children: React.ReactChild[] | React.ReactChild | undefined,
```
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/dashboard/src/components/Modal.tsx:5]: I'll revert. the "guard use case" which I used it for would also work with an empty container.

[the remark this answers is not in the document]
Comment (AlexTugarev) [reply components/dashboard/src/components/Modal.tsx]: done

File: components/dashboard/src/components/ContextMenu.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/dashboard/src/components/ContextMenu.tsx:67]: Oh, interesting. I thought the uniqueness was only required among sibling nodes

**Pull request merged**